libraw/src/decoder.rs

```diff
@@ -0,0 +1,38 @@
+use crate::{DecodedImage, Error, Result};
+use libraw_sys as sys;
+
+pub struct Decoder {
+    inner: *mut sys::libraw_data_t,
+}
+
+impl Decoder {
+    pub fn new() -> Self {
+        let inner = unsafe { sys::libraw_init(0) };
+        Self { inner }
+    }
+
+    pub fn decode(self, buf: &[u8]) -> Result<DecodedImage> {
+        Error::check(unsafe {
+            sys::libraw_open_buffer(self.inner, buf.as_ptr() as *const _, buf.len())
+        })?;
+        Error::check(unsafe { sys::libraw_unpack(self.inner) })?;
+        debug_assert!(!unsafe { (*self.inner).rawdata.raw_alloc }.is_null());
+
+        let decoded = DecodedImage::new(&unsafe { (*self.inner).rawdata }, &unsafe {
```

Comment:
I had `RawImage` (formerly `Decoded`) take ownership of `Processor`. I had to get the pointers before creation of the struct. I think that's fine but I'm skeptical because we get the `rawdata` pointer and create `Sizes` directly before moving Processor, which doesn't seem proper. However, it's using `Processor`'s `inner` (which I made public to the crate. Is that okay?) which is a pointer anyway so moving it shouldn't touch the data, so I think it's okay?

Reply:
Yup, the data is heap allocated, so it doesn't move when we move the pointer.